src/validation.cpp

```diff
@@ -1159,6 +1159,20 @@ CTransactionRef GetTransaction(const CBlockIndex* const block_index, const CTxMe
 {
     LOCK(cs_main);
 
+    if (mempool && !block_index) {
+        CTransactionRef ptx = mempool->get(hash);
+        if (ptx) return ptx;
+    }
+    if (g_txindex) {
+        CTransactionRef tx;
+        uint256 block_hash;
+        if (g_txindex->FindTx(hash, block_hash, tx)) {
+            if (!block_index || block_index->GetBlockHash() == block_hash) {
+                hashBlock = block_hash;
+                return tx;
+            }
```

Comment:
```suggestion
            }
            return nullptr;
```
Nit: This improves the performance of the (unlikely) case where a block hash is specified, but the transaction is found in a different block. The `if` at line 1170 would be false, so we would call `ReadBlockFromDisk()` below, but would not find the transaction there either (assuming the txindex is accurate).

I cherry-picked this PR with this suggestion onto #22437 as @jonatack [described above](https://github.com/bitcoin/bitcoin/pull/22383#issuecomment-879123985), and the tests all pass.

Reply:
How is this different from the initial solution: https://github.com/bitcoin/bitcoin/pull/22383#discussion_r663267393?

Reply:
@LarryRuane the idea is to allow fetching the transaction from a block out of the active chain - which -txindex doesn't take into account. Before this was possible.

Reply:
That could be checked by seeing if the specified block index is in the main chain.

But I suppose even that has edge cases (eg, index out of sync), so maybe it's better to just leave it alone.